### PR TITLE
Divergence-free Spectral Projection

### PR DESCRIPTION
**Pull request by vduruiss**
This is code for enforcing the divergence-free constraint in 2d vector field.

This is based on our [ICML Workshop Paper](https://openreview.net/forum?id=Zvxm14Rd1F)

I updated completely our earlier projection code to use an explicit formula for the linear least-squares problem for the divergence-free constraint, instead of keeping the memory-heavy and time-consuming pinv

@mliuschi  can also have a look

Note that the example and test are based on the new differentiation code from [PR643](https://github.com/neuraloperator/neuraloperator/pull/643)

### CODE REVIEW

File: neuralop/layers/spectral_projection.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+def spectral_projection_divergence_free(u, domain_size, constraint_modes):
+    """
+    Apply spectral projection to make velocity field divergence-free.
+    
+    This method implements a Helmholtz-Hodge projection in the spectral domain
+    to project velocity fields onto the divergence-free subspace.
+    
+    
+    References:
```
Comment (JeanKossaifi) [inline neuralop/layers/spectral_projection.py]: This is a really great doctoring!

Do all the sections actually get rendered? 
I would also put Parameters and Return first for consistency.

Comment (vduruiss) [reply neuralop/layers/spectral_projection.py]: The `plot_spectral_projection.py` example gets rendered correctly

Comment (vduruiss) [reply neuralop/layers/spectral_projection.py]: Fixed the docstring!

File: neuralop/layers/spectral_projection.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        neg_idx_h = torch.arange(height - constraint_modes[0]//2, height, 1)
+        idx_h = torch.cat((pos_idx_h, neg_idx_h)).to(device)
+        u_ft = torch.index_select(u_ft, 2, idx_h)
+    
+    # Extract lower width modes where the constraint is applied
+    if width != constraint_modes[1]:
+        pos_idx_w = torch.arange(0, constraint_modes[1]//2, 1)
+        neg_idx_w = torch.arange(width - constraint_modes[1]//2, width, 1)
+        idx_w = torch.cat((pos_idx_w, neg_idx_w)).to(device)
+        u_ft = torch.index_select(u_ft, 3, idx_w)
```
Comment (JeanKossaifi) [inline neuralop/layers/spectral_projection.py]: One option would be to index both dims jointly as we do in the main spectral conv, after an fftshift?

Comment (vduruiss) [reply neuralop/layers/spectral_projection.py]: Updated!

```  # Extract lower modes for both dimensions where the constraint is applied
    if (height != constraint_modes[0]) or (width != constraint_modes[1]):
        u_ft = torch.fft.fftshift(u_ft, dim=(2, 3))
        u_ft = u_ft[:, :, 
                    (height - constraint_modes[0])//2 : (height - constraint_modes[0])//2 + constraint_modes[0],
                    (width - constraint_modes[1])//2 : (width - constraint_modes[1])//2 + constraint_modes[1]]
        u_ft = torch.fft.ifftshift(u_ft, dim=(2, 3))
